### PR TITLE
Upgrade the project to Kotlin `1.6.10`.

### PR DESCRIPTION
**Pull request by vRallev**

### CODE REVIEW

File: buildSrc/src/main/java/Dependencies.kt [hunk 1]
```diff
@@ -19,10 +19,10 @@ object Dependencies {
...
     object Compose {
       const val activity = "androidx.activity:activity-compose:1.3.1"
-      const val foundation = "androidx.compose.foundation:foundation:1.0.1"
-      const val material = "androidx.compose.material:material:1.0.1"
-      const val tooling = "androidx.compose.ui:ui-tooling:1.0.1"
-      const val ui = "androidx.compose.ui:ui:1.0.1"
+      const val foundation = "androidx.compose.foundation:foundation:1.1.0-rc01"
+      const val material = "androidx.compose.material:material:1.1.0-rc01"
+      const val tooling = "androidx.compose.ui:ui-tooling:1.1.0-rc01"
+      const val ui = "androidx.compose.ui:ui:1.1.0-rc01"
```
Comment (vRallev) [inline buildSrc/src/main/java/Dependencies.kt:25]: Note that this is the latest version for these libraries, but there's an `rc02` for the compiler plugin with Kotlin 1.6.10 support. We'll use the same versions internally very soon.

File: buildSrc/src/main/java/Dependencies.kt [hunk 2]
```diff
@@ -60,15 +60,15 @@ object Dependencies {
   const val timber = "com.jakewharton.timber:timber:4.7.1"
 
   object Moshi {
-    const val adapters = "com.squareup.moshi:moshi-adapters:1.12.0"
-    const val codeGen = "com.squareup.moshi:moshi-kotlin-codegen:1.12.0"
-    const val moshi = "com.squareup.moshi:moshi:1.12.0"
+    const val adapters = "com.squareup.moshi:moshi-adapters:1.13.0"
+    const val codeGen = "com.squareup.moshi:moshi-kotlin-codegen:1.13.0"
+    const val moshi = "com.squareup.moshi:moshi:1.13.0"
```
Comment (vRallev) [inline buildSrc/src/main/java/Dependencies.kt:65]: To support Kotlin 1.6.10.

File: workflow-runtime/api/workflow-runtime.api [hunk 3]
```diff
@@ -80,3 +83,175 @@ public abstract interface class com/squareup/workflow1/WorkflowInterceptor$Workf
...
+	public static synthetic fun id$default (Lcom/squareup/workflow1/Workflow;Ljava/lang/String;ILjava/lang/Object;)Lcom/squareup/workflow1/internal/WorkflowNodeId;
+}
+
+public final class com/squareup/workflow1/internal/WorkflowRunner {
+	public fun <init> (Lkotlinx/coroutines/CoroutineScope;Lcom/squareup/workflow1/Workflow;Lkotlinx/coroutines/flow/StateFlow;Lcom/squareup/workflow1/TreeSnapshot;Lcom/squareup/workflow1/WorkflowInterceptor;)V
+	public final fun cancelRuntime (Ljava/util/concurrent/CancellationException;)V
+	public static synthetic fun cancelRuntime$default (Lcom/squareup/workflow1/internal/WorkflowRunner;Ljava/util/concurrent/CancellationException;ILjava/lang/Object;)V
+	public final fun nextOutput (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
+	public final fun nextRendering ()Lcom/squareup/workflow1/RenderingAndSnapshot;
+}
```
Comment (vRallev) [inline workflow-runtime/api/workflow-runtime.api:256]: The compiler starts spitting out so much more.

**Comment by rjrjr** [comment]
Is this a draft on purpose? Wondering if I should review yet.

File: buildSrc/src/main/java/Dependencies.kt (hunk 1 above)
Comment (rjrjr) [reply buildSrc/src/main/java/Dependencies.kt:25]: Could you [file an issue](https://github.com/square/workflow-kotlin/issues/new) tracking this, and add a comment here?

Comment (vRallev) [reply buildSrc/src/main/java/Dependencies.kt:25]: https://github.com/square/workflow-kotlin/issues/619

**Comment by vRallev** [comment]
> Is this a draft on purpose? Wondering if I should review yet.

I personally would wait until we use Kotlin 1.6 internally, otherwise upgrading Workflow internally is blocked. Hence I kept the PR in draft. You could review it now already, though.

File: samples/tictactoe/app/src/main/java/com/squareup/sample/mainactivity/TicTacToeComponent.kt [hunk 4]
```diff
@@ -17,13 +17,15 @@ import com.squareup.sample.gameworkflow.RealTakeTurnsWorkflow
...
 import com.squareup.sample.mainworkflow.TicTacToeWorkflow
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
 import io.reactivex.Single
 import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
 import timber.log.Timber
 
 /**
  * Pretend generated code of a pretend DI framework.
  */
+@WorkflowUiExperimentalApi
```
Comment (rjrjr) [inline samples/tictactoe/app/src/main/java/com/squareup/sample/mainactivity/TicTacToeComponent.kt:28]:
```suggestion
@OptIn(WorkflowUiExperimentalApi::class)
```

File: samples/tictactoe/common/src/main/java/com/squareup/sample/gameworkflow/RunGameWorkflow.kt [hunk 5]
```diff
@@ -42,6 +42,7 @@ typealias RunGameScreen = AlertContainerScreen<PanelContainerScreen<Any, Any>>
  * We define this otherwise redundant typealias to keep composite workflows
  * that build on [RunGameWorkflow] decoupled from it, for ease of testing.
  */
+@OptIn(WorkflowUiExperimentalApi::class)
```
Comment (rjrjr) [inline samples/tictactoe/common/src/main/java/com/squareup/sample/gameworkflow/RunGameWorkflow.kt:45]: I don't see why this is needed, possible b/c of the `@WorkflowUiExperimentalApi` mistake above? Same question for the other new appearances.

Comment (vRallev) [reply samples/tictactoe/common/src/main/java/com/squareup/sample/gameworkflow/RunGameWorkflow.kt:45]: The compiler is more strict. It's needed even with the other change.
```
> Task :samples:tictactoe:common:compileKotlin FAILED
e: /Users/ralf/Development/projects/square/workflow-kotlin/samples/tictactoe/common/src/main/java/com/squareup/sample/gameworkflow/RunGameWorkflow.kt: (46, 59): This declaration is experimental and its usage must be marked with '@com.squareup.workflow1.ui.WorkflowUiExperimentalApi' or '@OptIn(com.squareup.workflow1.ui.WorkflowUiExperimentalApi::class)'
```

**Review by vgonda — APPROVED** [review]
Seems fine to me

**Review by 0xMatthewGroves — APPROVED** [review]
LGTM.

**Comment by rjrjr** [comment]
Red shard is b/c request failures. Green API 21 shard is good enough for me, merging.

**Pull request merged**